File: packages/shared/src/market-indices.ts
/**
 * Índices de mercado — aliases + tipos (capa A discovery).
 * @see docs/engineering/lists-universes-design-2026-07-30.md
 */

export type MarketIndexHitDto = {
  code: string | null;
  displayName: string;
  yahooSymbol: string;
  region: string | null;
  currency: string | null;
  quoteType: string;
  source: string;
  constituentReady: boolean;
  score: number;
};

export type IndexConstituentMemberDto = {
  symbol: string;
  yahooSymbol: string;
  name: string | null;
};

export type IndexConstituentsDto = {
  indexCode: string;
  yahooIndexSymbol: string;
  provider: string;
  asOf: string;
  contentHash: string;
  members: IndexConstituentMemberDto[];
};

export type SubscribeMarketIndexResultDto = {
  listId: string;
  indexCode: string;
  displayName: string;
  yahooIndexSymbol: string;
  contentHash: string;
  instrumentIds: string[];
  progress: {
    total: number;
    alreadyPresent: number;
    imported: number;
    failed: string[];
    joined: string[];
    left: string[];
  };
  status: 'ready' | 'partial' | string;
};

export type CatalogIndexEntryDto = {
  code: string;
  displayName: string;
  yahooSymbol: string;
  region: string;
  currency: string;
  constituentReady: boolean;
  expectedCountMin: number;
  expectedCountMax: number;
  listId: string;
};

export type IndexSubscribeJobDto = {
  id: string;
  status: 'pending' | 'processing' | 'completed' | 'failed' | string;
  payload: {
    indexKey?: string;
    syncBars?: boolean;
    yearsBack?: number;
    [key: string]: unknown;
  };
  result: {
    phase?: string;
    checked?: number;
    total?: number;
    alreadyPresent?: number;
    imported?: number;
    failed?: string[];
    currentSymbol?: string;
    listId?: string;
    indexCode?: string;
    displayName?: string;
    instrumentIds?: string[];
    status?: string;
    joined?: string[];
    left?: string[];
    [key: string]: unknown;
  } | null;
  error: string | null;
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
};

/** Alias humano → código canónico (espejo Python). */
export const MARKET_INDEX_ALIASES: Record<string, string> = {
  ibex: 'IBEX35',
  ibex35: 'IBEX35',
  'ibex 35': 'IBEX35',
  '^ibex': 'IBEX35',
  sp500: 'SPX',
  spx: 'SPX',
  's&p500': 'SPX',
  's&p 500': 'SPX',
  '^gspc': 'SPX',
  sp100: 'OEX',
  's&p100': 'OEX',
  's&p 100': 'OEX',
  oex: 'OEX',
  dax: 'DAX',
  dax40: 'DAX',
  '^gdaxi': 'DAX',
  ndx: 'NDX',
  nasdaq100: 'NDX',
  'nasdaq 100': 'NDX',
  eurostoxx50: 'STOXX50E',
  'euro stoxx 50': 'STOXX50E',
  stoxx50: 'STOXX50E',
  dow: 'DJI',
  dowjones: 'DJI',
  'dow jones': 'DJI',
  djia: 'DJI',
  '^dji': 'DJI',
  dji: 'DJI',
  ftse: 'FTSE100',
  ftse100: 'FTSE100',
  'ftse 100': 'FTSE100',
  '^ftse': 'FTSE100',
  ftsemib: 'FTSEMIB',
  'ftse mib': 'FTSEMIB',
  mib: 'FTSEMIB',
  'hang seng': 'HSI',
  hangseng: 'HSI',
  hsi: 'HSI',
  '^hsi': 'HSI',
};

export function canonicalMarketIndexCode(query: string): string | null {
  const key = query.trim().toLowerCase().replace(/\s+/g, ' ');
  return MARKET_INDEX_ALIASES[key] ?? null;
}

/** ID estable de InstrumentList catalog para un índice (espejo Python). */
export function catalogListIdForIndex(code: string): string {
  const normalized = code.trim().toUpperCase();
  if (normalized === 'IBEX35') return 'ibex35';
  return `idx-${normalized.toLowerCase()}`;
}

File: apps/api-python/src/bolsa_api/schemas/market_indices.py
"""DTOs HTTP de índices y constitutivos."""

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class IndexHitDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    code: str | None = None

    display_name: str = Field(alias="displayName")

    yahoo_symbol: str = Field(alias="yahooSymbol")

    region: str | None = None

    currency: str | None = None

    quote_type: str = Field(alias="quoteType")

    source: str

    constituent_ready: bool = Field(alias="constituentReady")

    score: float = 0





class MarketIndicesSearchResponseDto(BaseModel):

    data: list[IndexHitDto]





class CatalogIndexEntryDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    code: str

    display_name: str = Field(alias="displayName")

    yahoo_symbol: str = Field(alias="yahooSymbol")

    region: str

    currency: str

    constituent_ready: bool = Field(alias="constituentReady")

    expected_count_min: int = Field(alias="expectedCountMin")

    expected_count_max: int = Field(alias="expectedCountMax")

    list_id: str = Field(alias="listId")





class MarketIndexCatalogResponseDto(BaseModel):

    data: list[CatalogIndexEntryDto]





class IndexConstituentMemberDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    symbol: str

    yahoo_symbol: str = Field(alias="yahooSymbol")

    name: str | None = None





class IndexConstituentDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    index_code: str = Field(alias="indexCode")

    yahoo_index_symbol: str = Field(alias="yahooIndexSymbol")

    provider: str

    as_of: str = Field(alias="asOf")

    content_hash: str = Field(alias="contentHash")

    members: list[IndexConstituentMemberDto]





class IndexConstituentsResponseDto(BaseModel):

    data: IndexConstituentDto





class SubscribeMarketIndexRequestDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    index_key: str = Field(alias="indexKey", min_length=1, max_length=40)

    sync_bars: bool = Field(default=False, alias="syncBars")

    years_back: int = Field(default=2, alias="yearsBack", ge=1, le=25)





class SubscribeProgressDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    total: int

    already_present: int = Field(alias="alreadyPresent")

    imported: int

    failed: list[str] = Field(default_factory=list)

    joined: list[str] = Field(default_factory=list)

    left: list[str] = Field(default_factory=list)





class SubscribeMarketIndexResultDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    list_id: str = Field(alias="listId")

    index_code: str = Field(alias="indexCode")

    display_name: str = Field(alias="displayName")

    yahoo_index_symbol: str = Field(alias="yahooIndexSymbol")

    content_hash: str = Field(alias="contentHash")

    instrument_ids: list[str] = Field(alias="instrumentIds")

    progress: SubscribeProgressDto

    status: str





class SubscribeMarketIndexResponseDto(BaseModel):

    data: SubscribeMarketIndexResultDto





class IndexSubscribeJobDto(BaseModel):

    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)  # type: ignore[typeddict-unknown-key]



    id: str

    status: str

    payload: dict[str, Any]

    result: dict[str, Any] | None = None

    error: str | None = None

    created_at: str = Field(alias="createdAt")

    updated_at: str = Field(alias="updatedAt")

    completed_at: str | None = Field(alias="completedAt", default=None)





class IndexSubscribeJobResponseDto(BaseModel):

    data: IndexSubscribeJobDto


